Update EIP-1: EIP pain relief (#7230)
ethereum/EIPs: EIPS/eip-1.md

Author: Yeah, but too late now.  My apologies.

status: Living
type: Meta
author: Martin Becze <mb@ethereum.org>, Hudson Jameson <hudson@ethereum.org>, et al.

> discussions-to: https://ethereum-magicians.org/t/update-eip-1-eip-pain-relief-7230/15082

Reviewer: Is this meant to be here?

Reviewer 2: genial laborioso

## EIP Types

There are three types of EIP: Standards Track, Meta and Informational.

### Standards Track EIP

> Describes any change that affects most or all Ethereum implementations, such as—a change to the network protocol, a change in block or transaction validity rules, proposed application standards/conventions, or any change or addition that affects the interoperability of applications using Ethereum. Standards Track EIPs consist of three parts—a design document, an implementation, and (if warranted) an update to the [formal specification](https://github.com/ethereum/yellowpaper). Furthermore, Standards Track EIPs can be broken down into the following categories:

Reviewer 3: ```suggestion
Describes any change that affects most or all Ethereum implementations, such as—a change to the network protocol, a change in block or transaction validity rules, proposed application standards/conventions, or any change or addition that affects the interoperability of applications using Ethereum. Standards Track EIPs consist of three parts—a design document, an implementation, and (if warranted) an update to the [formal specification](https://github.com/ethereum/execution-specs). Furthermore, Standards Track EIPs can be broken down into the following categories:
```
😉

#### Core

Improvements requiring a consensus fork (e.g. [EIP-5](./eip-5.md), [EIP-101](./eip-101.md)), as well as changes that are not necessarily consensus critical but may be relevant to [“core dev” discussions](https://github.com/ethereum/pm) (for example, [EIP-90], and the miner/node strategy changes 2, 3, and 4 of [EIP-86](./eip-86.md)).
]

> #### Networking

Reviewer 3: Didn't you delete this and interface eips above?

Includes improvements around [devp2p](https://github.com/ethereum/devp2p/blob/readme-spec-links/rlpx.md) ([EIP-8](./eip-8.md)) and [Light Ethereum Subprotocol](https://ethereum.org/en/developers/docs/nodes-and-clients/#light-node), as well as proposed improvements to network protocol specifications of [whisper](https://github.com/ethereum/go-ethereum/issues/16013#issuecomment-364639309) and [swarm](https://github.com/ethereum/go-ethereum/pull/2959).

#### Interface

> Includes improvements around client [API/RPC](https://github.com/ethereum/execution-apis#README) specifications and standards, and also certain language-level standards like method names ([EIP-6](./eip-6.md)) and [contract ABIs](https://docs.soliditylang.org/en/develop/abi-spec.html). The label “interface” aligns with the [interfaces repo] and discussion should primarily occur in that repository before an EIP is submitted to the EIPs repository.

Author: I'm not sure, but I'm not proposing any changes here.

## EIP Types

> There are three types of EIP: Standards Track, Meta and Informational.

Author: It might. It might also be worth pulling all of the EIPs together into one working group-- there aren't that many Networking and Interface proposals, they tend to be more closely related to Core work than ERCs, (some require client changes?) and the core devs are probably more competent than most to review them.

> There are three types of EIP: Standards Track, Meta and Informational.

Author: If we do that, Victor and I have considered creating an ERC working group, if he can find a few other people to help.  If we can work out how to set up repos for each group something close to a split could be realized without an actual bifurcation.

> There are three types of EIP: Standards Track, Meta and Informational.

Reviewer: I would be happy to join an ERC WG.

### General Requirements

It is highly recommended that a single EIP contain a single key proposal or new idea. The more focused the EIP, the more successful it tends to be. A change to one client doesn't require an EIP; a change that affects multiple clients, or defines a standard for multiple apps to use, does.

An EIP must meet certain minimum criteria. It must be a clear and complete description of the proposed enhancement. The enhancement must represent a net improvement. The proposed implementation, if applicable, must be solid and must not complicate the protocol unduly.

## Working Groups

Many proposals will come into EIP process via working groups who take responsibility for their specifications.  Working Group concerns include the technical quality, formatting and notations for their specifications, coordination with other specifications, the workflow for their tasks and documents, and addditional repos for tracking their work.

> * Working group proposals first enter the EIP process (see below) as **Drafts**.
> * While WG proposals are **Drafts** their Specification and workflow belongs to the WG.
> * WG EIPs reenter the EIP process to become **Final**.

Reviewer: ```suggestion
* Working group proposals are merged into the EIP repository with a status of `Draft`.
* As long as a working group EIP has a status of `Draft`, every member of the working group should have the same permissions as a co-author.
* With approval from an EIP editor, an EIP's working group can move that EIP's status from `Draft` to `Final`.
```
Is this wording better?

Author: A little better, but I really meant that "While WG proposals are **Drafts** their Specification and workflow belongs to the WG."  So it's up to the WG as to whether "every member of the working group should have the same permissions as a co-author."  Also up to the WG what it takes for them approve a PR from Draft to Final.  From there it's up to the Editors what our process is for accepting the PR, and I haven't changed that here.  I'd expect only minor editorial stuff would be appropriate at that point.

Reviewer: How about this then:

```
As long as a working group EIP has a status of `Draft`, the working group itself should be treated as if it were the sole author of the EIP for any process described in EIP-1 that requires author approval.
```

Author: Hmm.  I think if a WG wants to be a group author then they can without our help.  I can imagine scenarios where an author of an EIP makes a PR of their proposal to Final over the objections of the WG, and we merge it before becoming aware of the controversy, but don't see a need to spell that out here.  If we were to get into the nitty-gritty I'd suggest that at least one WG member should volunteer as an Editor who needs to approve PRs on that WG's EIPs.

> Currently, only the Core Developers are organized enough to already be doing this job.

Author: I'm willing to kick that can down the road until we have other people who want to be a working group -- by then we will have more experience.

As laid out here any group of people can self-organize as a working group and start vetting proposals according to whatever process and standards they like.  If other people prefer to submit proposals directly to the Editors that is fine too, but if the Editors judge that a proposal needs more technical review a relevant working group would be an obvious choice.  So a working group would succeed to the extent that people found them a useful means of creating EIPs.

The Core Devs are special in that nothing goes onto the mainnet unless they put it there, so there wouldn't be much point in trying to go around them.

### Core EIPs

> ***The Specification of Core EIPs is the responsibility of the Core Developers***, who the Editors treat as an independent Working Group.

Author: That would be fine.  Main issue is who would be the members of the groups.  The Core Devs already exist as a cohesive organization that wants more say over the Core EIP process, so I'm proposing to recognize that fact and give them more discretion.

## Working Groups

Many proposals will come into EIP process via working groups who take responsibility for their specifications.  Working Group concerns include the technical quality, formatting and notations for their specifications, coordination with other specifications, the workflow for their tasks and documents.  And, of course, their own governance.

* Working Group proposals first enter the EIP process (see below) as **Drafts**.
* While WG proposals are **Drafts** their Specification and workflow belongs to the WG.
* WG EIPs reenter the EIP process to become **Final**.

A Working Group fully governs its own EIPs.  To ensure this, at least one WG member should serve as an Editor with the power — within the governance of the WG — to merge PRs to EIPs/EIPS over the objections of the Github tooling or the other Editors.

Currently, only the Core Developers are organized enough to do this job.

### Core EIPs

***The Specification of Core EIPs is the responsibility of the Core Developers***, who the Editors treat as an independent Working Group.

### Editorial Services

> The Editors' job, for individual authors and for working groups, is to work with the authors to ensure that proposals merged to EIPs/EIPS meet standards of quality, format, style, and usage, and to provide administrative and publication services for EIPs and related digital assets.

Reviewer: ```suggestion
The editors' job, for individual authors and for working groups, is to work with the authors to ensure that proposals merged to EIPs/EIPS meet standards of quality, format, style, and usage, and to provide administrative and publication services for EIPs and related digital assets.
```
Editor here is not a proper noun

*In short, your role as the champion is to write the EIP using the style and format described below, shepherd the discussions in the appropriate forums, and build community consensus around the idea.*

### EIP Process

The following is the standardization process for all EIPs in all tracks:

![EIP Status Diagram](../assets/eip-1/EIP-process-update.jpg)

- **Idea** - An idea that is pre-draft. This is not tracked within the EIP Repository.
- **Draft** - The first formally tracked stage of an EIP in development. An EIP is merged by an EIP Editor into the EIP repository when it meets some initial requirements: It should be technically sound, the Title and Preamble must be correct, and all included Sections should be readable.  Problems with language like spelling, grammar, markup, etc. are only considered warnings.

> - **Review** - An EIP Author marks an EIP as ready for and requesting Peer Review.  An Editor will merge the EIP when it meets all EIP requirements: it should be a technically sound, complete specification, properly formatted, with correct language, markup and external references.

Reviewer: The wording here feels a bit off.

Author: Agreed.  Just a sketch to get discussion started.

- **Last Call** - This is the final review window for an EIP before moving to `Final`. An EIP editor will assign `Last Call` status and set a review end date (`last-call-deadline`), typically 14 days later.
  - If this period results in necessary normative changes it will revert the EIP to `Review`.
- **Final** - This EIP represents the final standard. A Final EIP exists in a state of finality and should only be updated to correct errata and add non-normative clarifications.

>   - A PR moving an EIP from Last Call to Final SHOULD contain no changes other than the status update. Any content or editorial proposed change SHOULD be separate from this status-updating PR and committed prior to it.

Reviewer: How about any status to any status?

Author: Not sure, but I don't think I changed this.

Reviewer 4: I am ok that any changing status shall be a separate PR @Pandapip1

Author: Actually, this might conflict with enforcing EIPW only at status changes.

Reviewer 4: I think there is value running EIPW on non-status PRs for the lines being changed.

Author: What has bitten me in the past is the rules changing in the course of drafting an EIP.  I'd much rather see warnings that I can plan to deal with later rather than hard errors that force me to deal with them right away.  Once it comes time to move to the next phase then indeed it's time to get everything up to the latest standards.

Reviewer 5: huge +1 to "warning not error/merge-blocker"

  - *EIP Authors are notified of any algorithmic change to the status of their EIP*
- **Withdrawn** - The EIP Author(s) have withdrawn the proposed EIP. This state has finality and can no longer be resurrected using this EIP number. If the idea is pursued at later date it is considered a new proposal.
- **Living** - A special status for EIPs that are designed to be continually updated and not reach a state of finality. This includes most notably EIP-1.

### ERC Peers

Core EIPs are intensely scrutinized by the client teams and other developers and EIP Editors are generally familiar with the Ethereum protocol, whereas ERCs cover a wide and growing range of applications and are often outside the expertise of any Editor.  Historically, this has put a strain on the Editors' responsibility to maintain quality.  Therefore, an ERC Author should, with the help of the Editors, designate at least two Peers with relevant expertise who can help the Author refine the Idea technically, edit Drafts, and generally support the Author through the stages of the EIP Process.

## What belongs in a successful EIP?

Each EIP should have the following parts:

> - **Preamble** - RFC 822 style headers containing metadata about the EIP, including the EIP number, a short descriptive title (limited to a maximum of 44 characters), a description (limited to a maximum of 140 characters), and the author details. Irrespective of the category, the title and description should not include EIP number. See [below](./eip-1.md#eip-header-preamble) for details.

Reviewer: I feel like these should be formatted as Nth level headings instead of bullets.

Author: I'll try it.

*In short, your role as the champion is to write the EIP using the style and format described below, shepherd the discussions in the appropriate forums, and build community consensus around the idea.*

### EIP Process

The following is the standardization process for all EIPs in all tracks:

![EIP Status Diagram](../assets/eip-1/EIP-process-update.jpg)

#### Idea

An idea that is pre-draft. This is not tracked within the EIP Repository.

#### Draft

> The first formally tracked stage of an EIP in development. An EIP is merged by an EIP Editor into the EIP repository when it meets some initial requirements: It should be technically sound, the Title and Preamble must be correct, and all included Sections should be readable.  Problems with language like spelling, grammar, markup, etc. are only considered warnings.

Reviewer 6: I would drop the "technically sound" requirement, and expand a bit:

```suggestion
Draft is the first formally tracked stage of an EIP in development. An EIP is merged by an EIP Editor into the EIP repository when it meets some initial requirements: the title and preamble must be complete and descriptive, and required sections must be present and readable (but not necessarily complete.)  Problems with language like spelling, grammar, etc. are only considered warnings.

The specification itself does not need to be completely fleshed out, but should contain enough content to give a good idea of the intent of the proposal. Importantly, the proposal must be a good fit for the EIP process, as defined by the EIP Editors.
```

Author: I didn't change the "technically sound" part, and am inclined to keep it.  Otherwise I'm fine with your suggesting.

Author: Also, I'm not sure what "a good fit for the EIP process" means.  I wrote "must actually be a proposal to improve Ethereum", but I'm not sure if that's better.

Reviewer 6: I'd like to include the rule somewhere that editors can reject things that have little business being standards.

For example, we might reject:

 * `EIP-XXXX: Solidity Style Guide` because it isn't a technical specification, 
 * `EIP-XXXX: An implementation of secp256r1` because it doesn't define how two or more parties interoperate, or
 * `EIP-XXXX: A comprehensive table of precompile addresses` because it can never go to final.

I'm not sure how to write that concisely without unduly limiting editor discretion though.

Reviewer 5: ^ I tend to agree with Sam's point, that:
- never being finalizable 
- not improving interop at any level of the EVM stack (incl wallet-to-wallet, for ex)
and
- not improving the functioning of clients
should all be valid reasons for saying no.  i defer to the group where/how best to insert these restrictions, but that certainly seems a good starting point for acceptance criteria.

Sidenote: by both *my*  version of Sam's three rules and his, an INFORMATIONAL EIP defining EVM "compatibility" could be debatably in or out of scope.  That implies to me that the scope of "interop" should perhaps be more explicit, so it's less subjective whether defining EVM (at least for 2023) is a good EIP document or not :D

Author: I agree with where this is going, but would rather not resolve it just yet.

#### Review

> An EIP Author marks an EIP as ready for and requesting Peer Review.  An Editor will merge the EIP when it meets all EIP requirements: it should be a technically sound, complete specification, properly formatted, with correct language, markup and external references.

Reviewer 6: ```suggestion
An EIP Author marks an EIP as ready for and requesting Peer Review.  An Editor will merge the status change when the proposal meets all EIP requirements: it should be a complete specification, properly formatted, with correct language, markup and external references.
```

Author: Still wanting to keep "technically sound".

Reviewer 6: I'm not qualified to judge technical soundness for all EIPs. Maybe we go with a lower bar of plausibly sound?

Reviewer 5: Yeah, Technical Assistants could help editors feel qualified/informed enough to make this assessment, but delegating that assesment to the consensus of a WG (however tightly or loosely defined) of, say, 2 or more competing implementers and 1 or more editors-or-technical-assistants sounds like a great option if it's on the table!

Author: I'm resolving this unchanged for now.  Outside review is helpful and often necessary, but in the end the Editors have to make the call.

#### Last Call

> This is the final review window for an EIP before moving to `Final`. An EIP editor will assign `Last Call` status and set a review end date (`last-call-deadline`), typically 14 days later.
> 
> *If this period results in necessary normative changes it will revert the EIP to `Review`.*

Reviewer 6: ```suggestion
This is the final review window for an EIP before moving to `Final`. An EIP Author may request `Last Call` status and set a review end date (`last-call-deadline`), typically 14 days later.

*If this period results in necessary normative changes it will revert the EIP to `Review`.*
```

Author: I'm not sure why you want to change this.  The Author may request Last Call, but it seems like an Editors job to make the move.  But I'm not  that concerned either way.

Reviewer 6: I've always operated under the assumption that authors request a status change, and editors grant it.

I'm concerned that the wording "[a]n EIP editor will assign `Last Call` status" makes it sound like authors don't need to open a pull request. Without a pull request, I don't see how a proposal can move to last call.

Reviewer 5: if part of the problem is public perception of editorial omnipotence, it never hurts to *phrase* things less editor-centrically, as an author-push not an editor-pull :D

#### Final

This EIP represents the final standard. A Final EIP exists in a state of finality and should only be updated to correct errata and add non-normative clarifications.

A PR moving an EIP from Last Call to Final SHOULD contain no changes other than the status update. Any content or editorial proposed change SHOULD be separate from this status-updating PR and committed prior to it.

#### Stagnant

> Any EIP in `Draft` or `Review` or `Last Call` if inactive for a period of 6 months or greater is moved to `Stagnant`. An EIP may be resurrected from this state by Authors or EIP Editors through moving it back to `Draft` or it's earlier status. If not resurrected, a proposal may stay forever in this status.

Reviewer 6: ```suggestion
Any EIP in `Draft` or `Review` or `Last Call` if inactive for a period of 6 months or greater is moved to `Stagnant`. An EIP may be resurrected from this state by Authors or EIP Editors through moving it back to `Draft` or its earlier status. If not resurrected, a proposal may stay forever in this status.
```

#### Reference Implementation *(optional)*

An optional section that contains a reference/example implementation that people can use to assist in understanding or implementing this specification. This section may be omitted for many EIPs.  Most ERCs should include an implementation. Final Core EIPs should link to the relevant Execution Client or Consensus Client commits that implement the EIP.

#### Security Considerations

All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life-cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.

> #### Normative References
> 
> **Informative References**
> 
> Complete references for each external resource must be included in one of these two sections, as specified below.  

Reviewer 6: Perhaps something like this:

```suggestion
#### References

**Normative References**

**Informative References**

Complete references for each external resource must be included in one of these two sections, as specified below.  
```

*In short, your role as the champion is to write the EIP using the style and format described below, shepherd the discussions in the appropriate forums, and build community consensus around the idea.*

### EIP Process

The following is the standardization process for all EIPs in all tracks:

![EIP Status Diagram](../assets/eip-1/EIP-process-update.jpg)

#### Idea

An idea that is pre-draft. This is not tracked within the EIP Repository.

> #### Draft

Author: It makes sense, but outside of Core EIPs that hasn't really happened.  I'm proposing Technical Assistants (or the like) as a workable alternative.

The specification itself does not need to be completely fleshed out, but should contain enough content to give a good idea of the intent of the proposal. Importantly, the proposal must actually be a proposal to improve Ethereum.

#### Review

An EIP Author marks an EIP as ready for and requesting Peer Review.  An Editor will merge the EIP when it meets all EIP requirements: it should be a technically sound, complete specification, properly formatted, with correct language, markup and external references.

#### Last Call

This is the final review window for an EIP before moving to `Final`. An EIP editor will assign `Last Call` status and set a review end date (`last-call-deadline`), typically 14 days later.

*If this period results in necessary normative changes it will revert the EIP to `Review`.*

#### Final

> This EIP represents the final standard. A Final EIP exists in a state of finality and should only be updated to correct errata and add non-normative clarifications.

Author: Yeah, we've been missing an Errata section for a long time.

## What belongs in a successful EIP?

### EIP Sections

Each EIP should have the following sections:

#### Preamble

RFC 822 style headers containing metadata about the EIP, including the EIP number, a short descriptive title (limited to a maximum of 44 characters), a description (limited to a maximum of 140 characters), and the author details. Irrespective of the category, the title and description should not include EIP number. See [below](./eip-1.md#eip-header-preamble) for details.

#### Abstract

A multi-sentence (short paragraph) technical summary. This should be a very terse and human-readable version of the specification section. Someone should be able to read only the abstract to get the gist of what this specification does.

#### Motivation *(optional)*

> A motivation section is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. This section may be omitted if the motivation is evident.

Reviewer 7: Would it be useful for these two to be merged? In the EIPs I've read through often times they re-iterate the same information for the most part.

Author: It might sometimes, but I'm not proposing a change here.  (The Abstract really should be brief, whereas the Motivation can take as much space as it needs.)

#### Final

This EIP represents the final standard. A Final EIP exists in a state of finality and should only be updated to correct errata and add non-normative clarifications.

A PR moving an EIP from Last Call to Final SHOULD contain no changes other than the status update. Any content or editorial proposed change SHOULD be separate from this status-updating PR and committed prior to it.

#### Stagnant

Any EIP in `Draft` or `Review` or `Last Call` if inactive for a period of 6 months or greater is moved to `Stagnant`. An EIP may be resurrected from this state by Authors or EIP Editors through moving it back to `Draft` or it's earlier status. If not resurrected, a proposal may stay forever in this status.

*EIP Authors are notified of any algorithmic change to the status of their EIP*

#### Withdrawn

The EIP Author(s) have withdrawn the proposed EIP. This state has finality and can no longer be resurrected using this EIP number. If the idea is pursued at later date it is considered a new proposal.

#### Living

A special status for EIPs that are designed to be continually updated and not reach a state of finality. This includes most notably EIP-1.

### Technical Assistants

> Core EIPs are intensely scrutinized by the client teams and other developers for technical soundness, but the EIP Editors may still lack relevant expertise. And the numerous ERCs cover a wide and growing range of applications that are even more likely to be outside the expertise of any Editor.  Historically, this has put a strain on the Editors' responsibility to maintain quality.  Therefore, the Editors *at their discretion* may require Technical Assistants with the relevant expertise who can review the Author's Idea and work with the Author to ensure a technically sound proposal.

Reviewer 3: This is a good idea, but I think we've tried on a few occasions to activate these people, but have failed.

Author: @xinbenlv think we need make it a solid policy and try again. We've also failed to get enough new Editors to handle all of the proposals we get.  I think part of the difficulty is that proposals span too wide of a range of technical expertise for a small group of Editors.  Many publishers handle this by weeding out articles via peer review before editing even starts.

Reviewer 5: some kind of nominal honorarium or [on-chain?] badge of honor for participating never hurts to make folks feel appreciated for this "unpaid overtime".  i also think there's a cultural change where having engineers from your *competitors*  reviewing and approving your EIPs could really be a prestige/maturity symbol! in any case, I am a huge +1 to formalizing recognition of review and getting external/horizontal/non-editor review.

Author: Expecting laurels for service to Ethereum?!

Reviewer 5: > This is a good idea, but I think we've tried on a few occasions to activate these people, but have failed.

I was not paying attention when these occasions came and went, does anyone have links to catch me up on these attempts?

## What belongs in a successful EIP?

### EIP Sections

> Each EIP should have the following sectio:

Reviewer 3: ```suggestion
Each EIP should have the following sections:
```

#### Withdrawn

The EIP Author(s) have withdrawn the proposed EIP. This state has finality and can no longer be resurrected using this EIP number. If the idea is pursued at later date it is considered a new proposal.

#### Living

A special status for EIPs that are designed to be continually updated and not reach a state of finality. This includes most notably EIP-1.

### Technical Peers

Core EIPs are intensely scrutinized by the client teams and other developers for technical soundness, but the EIP Editors may still lack relevant expertise. And the numerous ERCs cover a wide and growing range of applications that are even more likely to be outside the expertise of any Editor.  Historically, this has put a strain on the Editors' responsibility to maintain quality.  Therefore, the Editors *at their discretion* may designate Technical Peers with the relevant expertise who can work with the Author to ensure a technically sound proposal.

## What belongs in a successful EIP?

### EIP Sections

*Note that Working Group EIPs may have additional sections at the WG's discretion, as may any EIP at the Editors' discretion.*

Each EIP should have the following sections:

#### Preamble

> RFC 822 style headers containing metadata about the EIP, including the EIP number, a short descriptive title (limited to a maximum of 44 characters), a description (limited to a maximum of 140 characters), and the author details. Irrespective of the category, the title and description should not include EIP number. See [below](./eip-1.md#eip-header-preamble) for details.

Reviewer: ```suggestion
RFC 822 style headers containing metadata about the EIP, including the EIP number, a short descriptive title (limited to a maximum of 44 characters), a description (limited to a maximum of 140 characters), and the author details. Irrespective of the category, the title and description should not include the EIP number. See [below](#eip-header-preamble) for details.
```

#### Abstract

A multi-sentence (short paragraph) technical summary. This should be a very terse and human-readable version of the specification section. Someone should be able to read only the abstract to get the gist of what this specification does.

> #### Motivation *(optional)*

Reviewer: ```suggestion
#### Motivation

This section is optional.
```

A motivation section is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. This section may be omitted if the motivation is evident.

#### Specification

The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (besu, erigon, ethereumjs, go-ethereum, nethermind, or others).

Working groups MAY specify additional requirements, including on the form of the Specification and the implementation of the technology.

> #### Rationale *(optional)*

Reviewer: ```suggestion
#### Rationale

This section is optional.
```

The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale should discuss important objections or concerns raised during discussion around the EIP.

> #### Backwards Compatibility *(optional)*

Reviewer: ```suggestion
#### Backwards Compatibility

This section is optional for all EIPs except those with a status of `Last Call` or `Final`.
```

All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their consequences. The EIP must explain how the author proposes to deal with these incompatibilities. This section may be omitted if the proposal does not introduce any backwards incompatibilities, but this section must be included if backward incompatibilities exist.

> #### Test Cases *(optional)*

Reviewer: ```suggestion
#### Test Cases

This section is optional for all non-Core EIPs.
```

Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Tests should either be inlined in the EIP as data (such as input/expected output pairs, or included in `../assets/eip-###/<filename>`. This section may be omitted for non-Core proposals.

> #### Reference Implementation *(optional)*

Reviewer: ```suggestion
#### Reference Implementation

This section is optional.
```

An optional section that contains a reference/example implementation that people can use to assist in understanding or implementing this specification. This section may be omitted for many EIPs.  Most ERCs should include an implementation. Final Core EIPs should link to the relevant Execution Client or Consensus Client commits that implement the EIP.

#### Security Considerations

All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life-cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.

> #### References *(if any)*

Reviewer: ```suggestion
#### References

This section is optional if there are no references.
```

*In short, your role as the champion is to write the EIP using the style and format described below, shepherd the discussions in the appropriate forums, and build community consensus around the idea.*

### EIP Process

> The following is the full, default standardization process for all EIPs in all tracks.  The diagram shows all of the  stages a document can be in and the transitions beteen those states. 

Reviewer: ```suggestion
The following is the full, default standardization process for all EIPs in all tracks. The diagram shows all of the  stages a document can be in and the transitions between those states. 
```
Typos

![EIP Status Diagram](../assets/eip-1/EIP-process-update.jpg)

> ***For Working Groups only Draft, Final, and Withdrawn are used in the Preamble as a `status`.  While a document is a Draft the working group is responsible for its process, which the group can track as `working-group-status`.***

Reviewer: ```suggestion
***For Working Groups only Draft, Final, and Withdrawn are used in the Preamble as a `status`. While a document is a Draft the working group is responsible for its process, which the group can track as `working-group-status`.***
```

> All other proposals will move through all stages of this process.

Reviewer: ```suggestion
All other proposals can move through all stages of this process.
```

#### `type` header

The `type` header specifies the type of EIP: Standards Track, Meta, or Informational. If the track is Standards please include the subcategory (core, networking, interface, or ERC).

#### `category` header

The `category` header specifies the EIP's category. This is required for standards-track EIPs only.

#### `created` header

The `created` header records the date that the EIP was assigned a number. Both headers should be in yyyy-mm-dd format, e.g. 2001-08-14.

#### `requires` header

EIPs may have a `requires` header, indicating the EIP numbers that this EIP depends on. If such a dependency exists, this field is required.

A `requires` dependency is created when the current EIP cannot be understood or implemented without a concept or technical element from another EIP. Merely mentioning another EIP does not necessarily create such a dependency.

### Linking to External Resources

Other than the specific exceptions listed below, links to external resources **SHOULD NOT** be included. External resources may disappear, move, or change unexpectedly.

The process governing permitted external resources is described in [EIP-5757](./eip-5757.md).

> External resources not so permitted **MAY** be included *at the Editors' discretion*.
> 

Reviewer 6: This is already permitted by the "**SHOULD NOT**" above.

```suggestion
```

Author: Hmm.  I think the "MAY" mostly eases the tone of the "SHOULD NOT", so not completely redundant.

A complete reference to every external resource **MUST** be included in a Normative or Informative References section.

> We take several precautions to ensure the continued accessibility of all external resources.  First, we follow the requirements of the [RFC 7322 RFC Style Guide](https://www.rfc-editor.org/rfc/rfc7322.html): references **MUST** be full citations, including author(s), title, and publication information -- URLs are acceptable, but **MUST NOT** be the sole information provided for a reference.  In addition, Links **SHOULD** meet the Requirements for Origins in [EIP-5757](./eip-5757.md), and a Digital Object Identifier (DOI), where available, **MUST** be included. Within "fair use", short quotations from referenced works are encouraged, as embedded quotations will not change. 

Reviewer 6: I'd much prefer we use a computer tractable citation format.

There is an example of how `csl-json` looks when rendered in [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963#fn:rfc4122), and [the relevant source](https://github.com/ethereum/EIPs/blob/e4cb6fda2825de5e8afbe27d33789f9558faaf8d/EIPS/eip-6963.md?plain=1#L248-L280).

You can convert between several well known formats (like BibTeX) here: https://citation.js.org/demo/ (choose CSL-JSON as output format.)

> URLs are acceptable, but **MUST NOT** be the sole information provided for a reference

Why not? If the URL contains a git commit hash or points to an exact document (like links to RFCs), I don't see why the URL alone is insufficient. We're an internet-first standards body. If anyone prints EIPs, I'd be quite surprised.

Author: I'd also prefer a computer-tractable citation format, but didn't try to specify one here.  For the URLs which we have "blessed" we can probably generate citations automatically.
 
 I took need for a non-URL reference from the IETF, which is the original internet-first standards body, and they apply the rule to their own standards.  As ever, the Editors can make exceptions if there is no other choice, but most documents at least have authors and titles.   Given that (and any other publication info) it is much easier to find a copy if the link goes dead.  And I don't see requiring an URL where a document is published but doesn't have a stable URL.  But I'm not a stickler on details here -- just the policy of "as complete a citation as possible."

Reviewer 6: I feel pretty strongly that someone with only an internet connection and a web browser should be able to view an EIP and all documents referenced therein.

Reviewer 7: For reference, CSL did create some confusion so I moved the linking to just the RFC link here. I like the way the CSL is getting rendered, but the way the IETF handles normative referencing seems to also work. If there's a way to easily generate a CSL-JSON object from a URL that would be ideal. I couldn't figure this out with the demo link provided above @SamWilsn.

Reviewer 6: https://citation.js.org/demo/ is how I format it. Doesn't automatically support pulling from a URL, but you can get, for example, bibtex and convert it to CSL-JSON.

Author: > I feel pretty strongly that someone with only an internet connection and a web browser should be able to view an EIP and all documents referenced therein.

I agree, they should, especially for normative links.  They should have enough information  to find the document in a library, and for a successful internet search if the link goes dead.

> According to the IESG Statement on [Normative and Informative References](https://www.ietf.org/about/groups/iesg/statements/normative-informative-references/), Normative references "specify documents that must be read to understand or implement the technology" whereas an Informative reference "only provides additional information".  The Specification section of an EIP **MUST NOT** contain Informative references, and **SHOULD** contain all necessary Normative references. Motivation and Rationale sections **MAY** include Informative references.

Reviewer 6: We do already have this distinction for EIPs (the `requires` preamble field.)

If the Specification section can only contain normative references, then I think we can infer that references in that section are normative. Similarly, if normative references are not permitted in non-Specification sections, we can infer that all references in those section are informative.

My point being: why have specific sections for normative and informative references? We can let authors use normal markdown links (and `csl-json` footnote blobs) to create references, and their location within the document decides whether they're normative or informative.

Author: We don't need them, you are right.  I took the one part from IETF before I wrote this part and didn't notice that I'd made it obsolete.   (And thanks for all your reviews here!)

At least one author must use a GitHub username, in order to get notified on change requests and have the capability to approve or reject them.

> #### `discussions-to` header

Reviewer 7: I've seen a few RFCs that have been moved through the process utilizing alternative channels of communication such as Telegram and Discord essentially subverting the intent of this header. Would it be useful to allow this to link to where the group is actually communicating to be more conducive for open communication? For example, one I'd really like to see done is to be able to link to an alternative GH repository where a spec is actually iterated on. The value of this is that it allows for separate issue tracking to make sure changes are getting implemented in the spec and Async debate to occur. Furthermore, it can link to authoritative places of discussion such as a Discord server or a Zoom link where synchronous discussions regularly occur via the Readme so that the discussions can be more transparent.

If we move in this direction we can also iterate on this process further so that the concept of Rings can be more formalized (e.g. form a ETH-Core-Devs org to iterate on Core specs) such that it makes it more known where the peer review and the experts are actually communicating.

Author: I've also seen cases where the Magician's link points off to other places where discussion is happening.  Does that suffice?

Author: Still, I agree that I'd rather see the bulk of the discussion happen here or on the Magicians where the history can be maintained.

#### `type` header

The `type` header specifies the type of EIP: Standards Track, Meta, or Informational. If the track is Standards please include the subcategory (core, networking, interface, or ERC).

#### `category` header

The `category` header specifies the EIP's category. This is required for standards-track EIPs only.

#### `created` header

The `created` header records the date that the EIP was assigned a number. Both headers should be in yyyy-mm-dd format, e.g. 2001-08-14.

#### `requires` header

EIPs may have a `requires` header, indicating the EIP numbers that this EIP depends on. If such a dependency exists, this field is required.

A `requires` dependency is created when the current EIP cannot be understood or implemented without a concept or technical element from another EIP. Merely mentioning another EIP does not necessarily create such a dependency.

### Linking to External Resources

Other than the specific exceptions listed below, links to external resources **SHOULD NOT** be included. External resources may disappear, move, or change unexpectedly.

The process governing permitted external resources is described in [EIP-5757](./eip-5757.md).

> External resources not so permitted **MAY** be included *at the Editors' discretion*.

Reviewer: ```suggestion
Other external resources may be allowed at Editor discretion.
```

Which renders to:

<!--- markdownlint-capture --->

> <!--- markdownlint-disable code-block-style --->

Reviewer 6: Why the extra dash?

Author: The dashes are from an ill-advised global search-and-replace.  I've fixed them, but the bot, in it's infinite wisdom, suddenly decided that the DOI example -- which has been broken since 2022-- must now be fixed before I can check in my changes.  And since I am a second-class editor I can't override it.

Author: Following is the error that just popped up.  I cannot find any difference to explain why it suddenly complained.

Author: ```
This is a sentence with a footnote.[^1]

[^1]:
    ```csl-json
    {
 Check failure on line 500 in EIPS/eip-1.md
GitHub
Actions
/ Markdown Linter

Code block style [Expected: fenced; Actual: indented]

EIPS/eip-1.md:500 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
      "type": "article",
      "id": 1,
      "author": [
```

Reviewer: It's always complained about this.

Reviewer: Also, the markdownlint bot isn't a required check.

Author: OK, thanks.  It does seem to be generating bad output at the end of the document, and I don't know how I broke it.

    }
    ```

> <!--- markdownlint-restore --->

Reviewer: Is there a reason you changed this?

The current EIP editors are

* Alex Beregszaszi (@axic)
* Gavin John (@Pandapip1)
* Greg Colvin (@gcolvin)
* Matt Garnett (@lightclient)
* Sam Wilson (@SamWilsn)
* Zainan Victor Zhou (@xinbenlv)
* Gajinder Singh (@g11tech)

Emeritus EIP editors are

* Casey Detrio (@cdetrio)
* Hudson Jameson (@Souptacular)
* Martin Becze (@wanderer)
* Micah Zoltu (@MicahZoltu)
* Nick Johnson (@arachnid)
* Nick Savers (@nicksavers)
* Vitalik Buterin (@vbuterin)

If you would like to become an EIP editor, please check [EIP-5069](./eip-5069.md).

## EIP Editor Responsibilities

For each new EIP that comes in, an editor does the following:

* Read the EIP to check if it is ready: sound and complete. The ideas must make technical sense, even if they don't seem likely to get to final status.
* The title should accurately describe the content.
* Check the EIP for language (spelling, grammar, sentence structure, etc.), markup (GitHub flavored Markdown), code style

> If the EIP isn't ready, the editor will send it back to the author with instructions for revision and possibly a request for review by Technical Peers. 

Reviewer 8: ```suggestion
If the EIP isn't ready, the editor will send it back to the author with instructions for revision and possibly a request for review by Technical Peers.
```

Reviewer 7: +1 to the suggestion. Would it make sense to get more specific and recommend a particular WG that they get review from as well?

## Copyright

Copyright and related rights waived via [CC0](../LICENSE.md).

> ](https://github.com/ethereum/EIPs/pull/7230)

Reviewer 3: ```suggestion
```

- The title should accurately describe the content.
- Check the EIP for language (spelling, grammar, sentence structure, etc.), markup (GitHub flavored Markdown), code style

> If the EIP isn't ready, the editor will send it back to the author for revision, with specific instructions and possibly a request for review by Technical Peers. 

Reviewer: ```suggestion
If the EIP isn't ready, the editor will send it back to the author for revision, with specific instructions.
```
This is already encompassed by "with specific instructions."

Author: Well, formally it is, but a reminder seems in order.  I'll reconsider the wording.
